### ISSUE TITLE
Bogus Error: Cannot infer this placeholder of type "parameters" (no type class instance found)

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#11576]
#### Description of the problem

```coq
Global Set Asymmetric Patterns.
Inductive type := base | arrow (s d : type).
Fixpoint for_each_lhs_of_arrow (f : Type) (t : type) : Type
  := match t with
     | base => unit
     | arrow s d => f * @for_each_lhs_of_arrow f d
     end.
Inductive expr {var : Type} : type -> Type :=
| Var {t} (v : var) : expr t
| Abs {s d} (f : var -> expr d) : expr (arrow s d)
.
Class parameters := {}.
Record rep {p : parameters} := {}.
Axiom listZ_local : forall {p}, @rep p.
Axiom p : parameters.
Existing Instance p.
Axiom cmd : forall {p : parameters}, Type.
Axiom ltype : forall {p : parameters} {listZ : @rep p}, Type.
Axiom translate_cmd
  : forall {p : parameters}
           (e : expr (var:=@ltype p (@listZ_local p)) base),
    cmd.
Axiom admit : forall {T}, T.
Existing Class rep.
Fixpoint translate_func' (pv:=_) {t} (e : @expr ltype t)
  : for_each_lhs_of_arrow ltype t -> @cmd pv :=
  match e with
  | Abs base d f =>
    fun (args : _ * for_each_lhs_of_arrow _ d) =>
      translate_func' (f (fst args)) (snd args)
  | Var base v =>
    fun _ => translate_cmd (Var v)
  | _ => fun _ => admit
  end.
(* File "./bug_01.v", line 30, characters 30-31:
Error: Cannot infer this placeholder of type "parameters" (no type class
instance found).*)
```

This is absurd, because there's an `Existing Instance`.  Also, if I remove `Existing Class rep`, then it goes through fine.

cc @mattam82 

#### Coq Version

8.9.0 (versions 8.10 and newer raise anomalies, cf #11572)

**Comment by JasonGross** [comment]
Also replacing `(pv:=_)` with `(pv:parameters:=_)` makes it go through.

**Comment by JasonGross** [comment]
Here's a smaller example of this one:
```coq
Class Foo := foo : True.
Instance: Foo := I.
Definition bar (v:=_) (H : @foo v = @foo v) : True := I.
```

**Comment by JasonGross** [comment]
My guess is that the issue here is that the "available for typeclass resolution" flag is not being inherited appropriately when evars are unified

**Comment by Alizter** [comment]
Seems to be fixed in 8.14. Will add to test-suite.
